scrape/scrape.go

```diff
@@ -1569,18 +1572,39 @@ loop:
 		// Increment added even if there's an error so we correctly report the
 		// number of samples remaining after relabeling.
 		added++
-
+		exemplars = exemplars[:0] // Reset and reuse the exemplar queue.
 		for hasExemplar := p.Exemplar(&e); hasExemplar; hasExemplar = p.Exemplar(&e) {
 			if !e.HasTs {
 				e.Ts = t
 			}
+			exemplars = append(exemplars, e)
+			e = exemplar.Exemplar{} // Reset for next time round loop.
+		}
+		sort.Slice(exemplars, func(i, j int) bool {
+			if exemplars[i].Ts != exemplars[j].Ts {
```

Comment:
Let's treat exemplars with `e.HasTs == false` specially and always sort them as the most recent, and sort them among each other just by value and label hash but not by their Ts set above. (Will comment later why I think it will help.)

Reply:
Note: This approach is discarded after our IRL discussion.